Source/WebKit/UIProcess/Automation/BidiStorageAgent.cpp

```diff
@@ -0,0 +1,292 @@
+#include "WebPageProxy.h"
+#include "WebsiteDataStore.h"
+
+namespace WebKit {
+
+using namespace Inspector;
+
+WTF_MAKE_TZONE_ALLOCATED_IMPL(BidiStorageAgent);
+
+BidiStorageAgent::BidiStorageAgent(WebAutomationSession& session, BackendDispatcher& backendDispatcher)
+    : m_session(session)
+    , m_storageDomainDispatcher(BidiStorageBackendDispatcher::create(backendDispatcher, this))
+{
+}
+BidiStorageAgent::~BidiStorageAgent() = default;
+
+static Ref<Inspector::Protocol::BidiStorage::PartialCookie> buildObjectForCookie(const WebCore::Cookie& cookie)
+{
+    auto partialCookie = Inspector::Protocol::BidiStorage::PartialCookie::create()
+        .setValue(cookie.value)
+        .setName(cookie.name)
+        .setDomain(cookie.domain)
+        .release();
+
+    partialCookie->setPath(cookie.path);
+    partialCookie->setHttpOnly(cookie.httpOnly);
+    partialCookie->setSecure(cookie.secure);
+    return partialCookie;
+}
+
+static Ref<JSON::ArrayOf<Inspector::Protocol::BidiStorage::PartialCookie>> buildArrayForCookies(const Vector<WebCore::Cookie>& cookiesList)
+{
+    auto cookies = JSON::ArrayOf<Inspector::Protocol::BidiStorage::PartialCookie>::create();
+
+    for (const auto& cookie : cookiesList)
+        cookies->addItem(buildObjectForCookie(cookie));
+
+    return cookies;
+}
+
+
+static bool matchesCookieFilter(const WebCore::Cookie& cookie, const RefPtr<JSON::Object> optionalFilter)
+{
+    if (!optionalFilter)
+        return true;
+
+    String filterNameOptional = optionalFilter->getString("name"_s);
+    if (!filterNameOptional.isEmpty()) {
+        if (cookie.name != filterNameOptional)
+            return false;
+    }
+
+    auto filterValueOptional = optionalFilter->getString("value"_s);
+    if (!filterValueOptional.isEmpty()) {
+        if (cookie.value != filterValueOptional)
+            return false;
+    }
+
+    auto filterDomainOptional = optionalFilter->getString("domain"_s);
+    if (!filterDomainOptional.isEmpty()) {
+        if (cookie.domain != filterDomainOptional)
+            return false;
+    }
+
+    auto filterPathOptional = optionalFilter->getString("path"_s);
+    if (!filterPathOptional.isEmpty()) {
+        if (cookie.path != filterPathOptional)
+            return false;
+    }
+
+    auto filterHttpOnlyOptional = optionalFilter->getBoolean("httpOnly"_s);
+    if (filterHttpOnlyOptional) {
+        if (cookie.httpOnly != filterHttpOnlyOptional.value())
+            return false;
+    }
+
+    auto filterSecureOptional = optionalFilter->getBoolean("secure"_s);
+    if (filterSecureOptional) {
+        if (cookie.secure != filterSecureOptional.value())
+            return false;
+    }
+
+    return true;
+}
+
+Inspector::Protocol::ErrorStringOr<Ref<Inspector::Protocol::BidiStorage::PartitionKey>> BidiStorageAgent::makePartitionKey(RefPtr<JSON::Object> partitionDescriptor)
+{
+    auto partitionKey = Inspector::Protocol::BidiStorage::PartitionKey::create().release();
+    if (!partitionDescriptor)
```

Comment:
I believe this should return Invalid Argument if there isn't any partition descriptor?